### ISSUE TITLE
When resuming execution in the debugger, the debugged app should be given the focus

### ISSUE THREAD

**Issue by tristanlabelle** [swiftlang/vscode-swift#657]
**Is your feature request related to a problem? Please describe.**
When hitting a breakpoint and resuming execution, it's useful to show the debugged app again. Xcode and big Visual Studio do this.

**Describe the solution you'd like**
When resuming execution, the debugger should find the main window associated with the process and bring it to the front.

**Describe alternatives you've considered**
Alt+tabbing back to the app.

**Additional context**
Unknown if this would need to be in the DAP or if the Swift extension has enough info to do it itself. It does feel like more of an editor/UI feature than DAP functionality. This would be OS-specific logic.

**Comment by adam-fowler** [comment]
Not sure I have control over this. It might be something you need to add an issue to the VSCode repo. Have you also checked the settings to see if there is one related to this?

**Comment by tristanlabelle** [comment]
There's no setting I found for this in VS Code, and I saw the dotnet/C# debugger doesn't implement this either.

I assumed this feature would be implementable in the Swift extension, because the DAP itself shouldn't be involved in presentation concerns and this is specific to the debugger type since it might not apply to a webpage. The Swift extension is the next layer in the stack that could have hooks to query the debugged process ID, find the Window and show it.

If you have more background and believe this is rather a missing VS Code feature, I can also file an issue there.

**Comment by adam-fowler** [comment]
You are right it has nothing to do with the DAP and is a UI thing. Unfortunately I don't think VSCode provides a hook into when continue is pressed. The suggestion is generic enough to be applicable to all debuggers so an issue on microsoft/vscode seems like a good idea.

**Comment by tristanlabelle** [comment]
I filed the bug upstream, and while thinking about it, there would be enough extensibility already for the Swift extension to use `DebugSession.customRequest` send a custom DAP request to the `swift-lldb` adapter to flip a boolean "Show app on resume" flag, at which point `swift-lldb` could implement this functionality and give the app focus upon receiving "continue" requests, having received a green light to do presentation-related things. What do you think?

**Comment by adam-fowler** [comment]
The DAP `lldb-vscode` is a separate process from VSCode. I don't see how it can control the main app focus.

**Comment by adam-fowler** [comment]
Sorry I misunderstood what you were trying to do here, you want the debugged app to regain focus.

**Comment by tristanlabelle** [comment]
> Sorry I misunderstood what you were trying to do here, you want the debugged app to regain focus.
Correct!

I was thinking of making it configurable because this is a presentation-level feature and not all debugging scenarios would require this (for example some automated debugging tool wouldn't).

Reading through, CodeLLDB just [forwards to LLDB](https://github.com/vadimcn/codelldb/blob/master/adapter/lldb/src/sb/sbprocess.rs#L64) and eventually calls [Process::PrivateResume](https://github.com/llvm/llvm-project/blob/76041a45bbe3cd2b3b3acad46267f27815e6a652/lldb/source/Target/Process.cpp#L3095), which has a set of pre-resume listeners that could be useful for this.

I'll file the issue on LLDB then.

**Comment by tristanlabelle** [comment]
Looking into this further, the Swift extension could already implement this using a `DebugAdapterTracker`:
- Use `onDidSendMessage` to listen to the [Process event](https://microsoft.github.io/debug-adapter-protocol/specification#Events_Process) and store the `systemProcessId`.
- Use `onWillReceiveMessage` to listen for `continue` messages sent by VS Code, and perform platform-specific logic to find the the main window of the target process and give it the focus.

I think this would be preferable to implementing it in the DAP since it is a presentation-level feature.

**Comment by adam-fowler** [comment]
Yes you are right I could catch the continue event in a `DebugAdapterTracker`. I already use one of these to catch log messages. 

I have no idea what the platform specific code to bring an application to the foreground would be. A quick search found this https://github.com/sentialx/node-window-manager. But it says Linux support is WIP and it hasn't had an update in 4 years.

**Comment by matthewbastien** [comment]
This is something that would have to be done in lldb-dap if they wanted to implement it. We don't want to have platform specific code to bring windows into focus.
